linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4208,27 +4222,65 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
     } else if (mutable) {
       withDedicatedVar(RefinedType(declaredType))
     } else {
-      val refinedType = value.tpe
+      def computeRefinedType(): RefinedType = bindingName match {
+        case _ if value.tpe.isExact || declaredType == AnyType =>
+          value.tpe
+        case _: Binding.Local =>
+          value.tpe
+        case Binding.This =>
+          /* When binding to `this`, if the declared type is not `AnyType`,
```

Comment:
Why does the nullability thing only apply if we're not dealing with `AnyType`?

Reply:
I've expanded in the comment of the first case.